Mathlib/RingTheory/MvPowerSeries/LinearTopology.lean

```diff
@@ -0,0 +1,180 @@
+
+variable {σ : Type*} {R : Type*} [Ring R]
+
+/-- A power series `f` belongs to the twosided ideal `basis σ R ⟨J, d⟩`
+if and only if `coeff R e f ∈ J` for all `e ≤ d`. -/
+theorem mem_basis_iff {f : MvPowerSeries σ R} {Jd : TwoSidedIdeal R × (σ →₀ ℕ)} :
+    f ∈ basis σ R Jd ↔ ∀ e ≤ Jd.2, coeff R e f ∈ Jd.1 := by
+  simp [basis]
+
+/-- If `J ≤ K` and `e ≤ d`, then we have the inclusion of twosided ideals
+`basis σ R ⟨J, d⟩ ≤ basis σ R ⟨K, e,>`. -/
+theorem basis_le {Jd Ke : TwoSidedIdeal R × (σ →₀ ℕ)} (hJK : Jd.1 ≤ Ke.1) (hed : Ke.2 ≤ Jd.2) :
+    basis σ R Jd ≤ basis σ R Ke :=
+  fun _ ↦ forall_imp (fun _ h hue ↦ hJK (h (le_trans hue hed)))
+
+/-- `basis σ R ⟨J, d⟩ ≤ basis σ R ⟨K, e⟩` if and only if `J ≤ K` and `e ≤ d`. -/
+theorem basis_le_iff {J K : TwoSidedIdeal R} {d e : σ →₀ ℕ} (hK : K ≠ ⊤) :
+    basis σ R ⟨J, d⟩ ≤ basis σ R ⟨K, e⟩ ↔ J ≤ K ∧ e ≤ d := by
+  constructor
+  · simp only [basis, TwoSidedIdeal.le_iff, TwoSidedIdeal.coe_mk', setOf_subset_setOf]
+    intro h
+    by_contra h'
+    simp only [not_and_or] at h'
+    rcases h' with h' | h'
+    · simp only [← coe_subset_coe, Set.not_subset] at h'
+      obtain ⟨a, haJ, haK⟩ := h'
+      apply haK
+      specialize h (monomial R e a) _ e (le_refl e)
+      · intro e' he'
+        classical
+        rw [coeff_monomial]
+        split_ifs
+        · exact haJ
+        · apply zero_mem
+      rwa [coeff_monomial_same] at h
+    · simp only [← inf_eq_right] at h'
+      apply hK
+      rw [eq_top_iff]
+      intro a _
+      specialize h (monomial R e a) _
+      · intro e' he'
+        convert zero_mem J
+        apply coeff_monomial_ne
+        rintro ⟨rfl⟩
+        exact h' (right_eq_inf.mpr he').symm
+      · specialize h e (le_refl e)
+        rwa [coeff_monomial_same] at h
+  · rintro ⟨hJK, hed⟩
+    exact basis_le hJK hed
+
+variable [TopologicalSpace R]
+
+-- We endow MvPowerSeries σ R with the product topology.
+open WithPiTopology
+
+/- variable (σ R) in
+theorem ringSubgroupsBasis :
+    RingSubgroupsBasis (fun (Jd : {J : TwoSidedIdeal R | (J : Set R) ∈ 𝓝 0} × (σ →₀ ℕ))
+        ↦ (basis σ R ⟨Jd.1, Jd.2⟩).asIdeal.toAddSubgroup) where
+  inter Jd Ke := ⟨⟨⟨Jd.1 ⊓ Ke.1, Filter.inter_mem Jd.1.prop Ke.1.prop⟩, Jd.2 ⊔ Ke.2⟩, by
+    simp only [le_inf_iff]
+    exact ⟨basis_le inf_le_left le_sup_left, basis_le inf_le_right le_sup_right⟩⟩
+  mul Jd := ⟨Jd, fun f ↦ by
+    simp only [Submodule.coe_toAddSubgroup, mem_mul]
+    rintro ⟨x, hx, y, hy, rfl⟩
+    exact Ideal.mul_mem_left _ _ hy⟩
+  leftMul f Jd := ⟨Jd, fun g hg ↦ (basis σ R ⟨Jd.1, Jd.2⟩).mul_mem_left f g hg⟩
+  rightMul f Jd := ⟨Jd, fun g hg ↦ by
+    intro e he
+    simp only [Submodule.coe_toAddSubgroup, TwoSidedIdeal.coe_asIdeal,
+      mem_coe, sub_zero, mem_basis_iff] at hg ⊢
+    classical
+    rw [coeff_mul]
+    apply sum_mem
+    rintro ⟨i, j⟩ h
+    apply TwoSidedIdeal.mul_mem_right
+    apply hg i (le_trans ?_ he)
+    simp only [← Finset.mem_antidiagonal.mp h, le_self_add]⟩
+-/
```

Comment:
Don't forget to remove this before merging

Reply:
;-)